Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CancellationRate.hs

```diff
@@ -0,0 +1,112 @@
+
+ as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+{-# LANGUAGE DerivingVia #-}
+{-# OPTIONS_GHC -Wno-deprecations #-}
+
+module SharedLogic.CancellationRate where
+
+import qualified Domain.Types.MerchantOperatingCity as DMOC
+import qualified Domain.Types.Person as DP
+import Kernel.Prelude
+import qualified Kernel.Storage.Hedis as Redis
+import Kernel.Types.Error
+import Kernel.Types.Id
+import qualified Kernel.Types.SlidingWindowCounters as SWC
+import Kernel.Utils.Common
+import qualified Kernel.Utils.SlidingWindowCounters as SWC
+import qualified Storage.CachedQueries.Merchant.TransporterConfig as CQTC
+
+data CancellationRateData = CancellationRateData
+  { assignedCount :: Maybe Int,
+    cancelledCount :: Maybe Int,
+    cancellationRate :: Maybe Int,
+    windowSize :: Maybe Int
+  }
+  deriving (Show, Eq, Generic, ToJSON, FromJSON)
+
+mkRideAssignedKey :: Text -> Text
+mkRideAssignedKey driverId = "driver-offer:CancellationRate:ride-assigned-to-driverId:" <> driverId
+
+mkRideCancelledKey :: Text -> Text
+mkRideCancelledKey driverId = "driver-offer:CancellationRate:ride-cancelled-by-driverId:" <> driverId
+
+getWindowSize :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => Id DMOC.MerchantOperatingCity -> m Integer
+getWindowSize mocId = do
+  merchantConfig <- CQTC.getTransporterConfigFromDB mocId >>= fromMaybeM (TransporterConfigNotFound mocId.getId)
+  pure $ toInteger $ fromMaybe 7 merchantConfig.cancellationRateWindow
+
+incrementCancelledCount ::
+  ( Redis.HedisFlow m r,
+    EsqDBFlow m r,
+    CacheFlow m r
+  ) =>
+  Id DP.Person ->
+  m ()
+incrementCancelledCount driverId = Redis.withCrossAppRedis $ SWC.incrementWindowCount (mkRideCancelledKey driverId.getId) (SWC.SlidingWindowOptions 15 SWC.Days)
+
+incrementAssignedCount ::
+  ( Redis.HedisFlow m r,
+    EsqDBFlow m r,
+    CacheFlow m r
+  ) =>
+  Id DP.Person ->
+  m ()
+incrementAssignedCount driverId = Redis.withCrossAppRedis $ SWC.incrementWindowCount (mkRideAssignedKey driverId.getId) (SWC.SlidingWindowOptions 15 SWC.Days)
+
+getCancellationCount ::
+  ( Redis.HedisFlow m r,
+    EsqDBFlow m r,
+    CacheFlow m r,
+    FromJSON a
+  ) =>
+  Integer ->
+  Id DP.Person ->
+  m [a]
+getCancellationCount windowSize driverId = fmap catMaybes . Redis.withCrossAppRedis $ SWC.getCurrentWindowValuesUptoLast windowSize (mkRideCancelledKey driverId.getId) (SWC.SlidingWindowOptions 15 SWC.Days)
```

Comment:
don't hard code these window values, use the same `windowSize` thing instead of 15